File: tests/browser/unit/widgets/test_worker_show_opens_in_background.mjs
// ============================================================================
// test_worker_show_opens_in_background.mjs — fix04, session 6d19df41, node 4.149.
//
// THE OPERATOR'S REPORT (session 6d19df41): «Hey. Why are you open the documents
// right now? I set open my email». A ghost "Scarborough" errand commissioned its
// `documento` sheet (src `worker:1`), and the show came to the FRONT over the
// email flow he was in.
//
// THE MECHANISM: a worker-commissioned card still OPENS — placed in free space,
// data rendering as usual — but never TAKES FOCUS over what he is looking at.
// `sse.js` marks worker-src shows as `background:true`; both hosts honor it
// (desktop skips `_bringFront`, mobile Deck skips `_goTo`).
//
// This MOUNTS the real `sse.js` handler over the real `store.js` with a stub
// host and replays the three provenances: worker / flash / user-echo.
//
// Run: node tests/browser/unit/widgets/test_worker_show_opens_in_background.mjs
// ============================================================================
import assert from "node:assert/strict";

const mem = new Map([["hb_lang", "es"], ["hb_i18n_es", JSON.stringify({})]]);
globalThis.localStorage = { getItem: k => (mem.has(k) ? mem.get(k) : null),
                            setItem: (k, v) => mem.set(k, String(v)), removeItem: k => mem.delete(k) };
globalThis.fetch = async () => ({ ok: true, json: async () => ({}), text: async () => "" });
globalThis.document = { documentElement: { lang: "es", setAttribute(){}, style: { setProperty(){} } },
                        querySelectorAll: () => [], querySelector: () => null, addEventListener(){},
                        createElement: () => ({ style: {}, classList: { add(){}, remove(){} },
                                                appendChild(){}, setAttribute(){} }),
                        head: { appendChild(){} }, body: { appendChild(){} } };
globalThis.window = { addEventListener(){}, location: { href: "http://localhost/" } };
let sink = null;
globalThis.EventSource = class { constructor() { sink = this; } };

await import("../../../../frontend/app/core/store.js?v=2");
const { openSSE } = await import("../../../../frontend/app/services/sse.js?v=2");
const calls = [];
const desktop = { show(id, opts){ calls.push([id, opts]); }, close(){}, closeAll(){},
                  refreshData(){}, createWidget(){}, modifyWidget(){}, onDeleted(){},
                  showConfirm(){}, hideConfirm(){}, move(){}, resize(){}, fullscreen(){},
                  refreshRegistry(){} };
openSSE(desktop);
assert.ok(sink && typeof sink.onmessage === "function", "the SSE handler did not mount");
const push = (o) => sink.onmessage({ data: JSON.stringify(o) });

// ── 1. the ghost errand's show: opens, but in background ─────────────────────────────────────────────────
push({ kind: "widget", label: "show", id: "documento", src: "worker:1", data: { title: "Scarborough" } });
assert.equal(calls.length, 1, "a worker-commissioned show must still OPEN the card");
assert.equal(calls[0][0], "documento");
assert.equal(calls[0][1] && calls[0][1].background, true,
  "THE BUG: a worker show took focus over his email flow — it must open in background");
assert.equal(calls[0][1] && calls[0][1].data && calls[0][1].data.title, "Scarborough",
  "background must not drop the pushed data");

// ── 2. voice-ordered shows still come to the front ────────────────────────────────────────────────────────
for (const evt of [{ kind: "widget", label: "show", id: "mensajeria", src: "flash" },
                   { kind: "widget", label: "show", id: "mensajeria" }]) {
  calls.length = 0;
  push(evt);
  assert.equal(calls.length, 1, `a ${evt.src || "sourceless"} show must open the card`);
  assert.ok(!calls[0][1] || calls[0][1].background !== true,
    `a ${evt.src || "sourceless"} show is his own order — it must come to the front`);
}

// ── 3. the canvas echo stays silent (V2-261, unchanged) ───────────────────────────────────────────────────
calls.length = 0;
push({ kind: "widget", label: "show", id: "mensajeria", src: "user" });
assert.equal(calls.length, 0, "a canvas report is not an order — still ignored");

console.log("ok: worker shows open in background, his own shows front, echo silent");

File: tests/browser/unit/widgets/test_worker_show_opens_in_background.py
"""A worker-commissioned card opens WITHOUT taking focus (fix04, session 6d19df41).

The operator: «Hey. Why are you open the documents right now? I set open my
email». A ghost "Scarborough" errand's `documento` sheet (show src `worker:1`)
came to the FRONT over his email flow. The mechanism: `sse.js` marks
worker-src shows as `background:true`; desktop `show()` then places the card
without `_bringFront`, and mobile Deck mounts it without `_goTo`. His own
voice-ordered shows (flash/sourceless) still front; the canvas echo stays
silent (V2-261).
"""
from __future__ import annotations

import re
import shutil
import subprocess
from pathlib import Path

import pytest

SCRIPT = Path(__file__).with_name("test_worker_show_opens_in_background.mjs")
DESKTOP = Path("frontend/app/widgets/desktop.js")
DECK = Path("frontend/mobile/app/shell/Deck.js")
SSE = Path("frontend/app/services/sse.js")


@pytest.mark.skipif(shutil.which("node") is None, reason="node not installed")
def test_a_worker_show_opens_but_does_not_take_focus():
    r = subprocess.run(["node", str(SCRIPT)], capture_output=True, text=True, timeout=60)
    assert r.returncode == 0, (r.stdout or "") + (r.stderr or "")


def _method(src: str, start: str, end: str) -> str:
    return src[src.index(start):src.index(end)]


def test_desktop_show_never_fronts_a_background_card():
    """Every `_bringFront` inside desktop `show()` answers to `background`.

    A source test cannot say the card stays down — the .mjs mounts only the
    SSE host contract, not the canvas — so this pins the other half: no
    `_bringFront` in `show()` may run for a background card.
    """
    body = _method(DESKTOP.read_text(encoding="utf-8"),
                   "async show(rawId,", "async _showActivity(")
    calls = [l.strip() for l in body.splitlines() if "_bringFront(" in l]
    # the pointerdown REGISTRATION is not a front — it only says clicks front later, as always
    calls = [l for l in calls if "=>" not in l]
    assert len(calls) == 3, f"show() must front in exactly 3 places, found {calls}"
    assert all("if(!background)" in l.replace(" ", "") for l in calls), \
        f"every front inside show() must answer to `background`: {calls}"


def test_deck_show_never_switches_to_a_background_card():
    body = _method(DECK.read_text(encoding="utf-8"),
                   "let baseId, id, wq;", "close(id) {")
    assert "if (!background) this._goTo" in body
    bare = [l for l in body.splitlines()
            if "this._goTo(" in l and "if (!background)" not in l]
    assert not bare, f"Deck.show() switches cards past the guard: {bare}"


def test_worker_provenance_is_detected_by_prefix_not_a_list():
    """`worker:<anything>` is background — a new worker kind must not need a new rule."""
    src = SSE.read_text(encoding="utf-8")
    assert "/^worker:/" in src, "the gate must match the worker: prefix"
    assert "background: _bg" in src, "the verdict must travel to the host as `background`"
